drivers: usb: udc: add NuMaker series USBD controller driver

ccli8 opened the pull request:
This is to add UDC driver for Nuvoton NuMaker series USBD controller.

Verified samples include:
- zephyr/samples/subsys/usb/cdc_acm
- zephyr/samples/subsys/usb/console
- zephyr/samples/subsys/usb/hid-mouse
- zephyr/samples/subsys/usb/mass

Fixes: #74660

ccli8 committed:
drivers: usb: udc: add NuMaker series USBD controller driver

Add UDC driver for Nuvoton NuMaker series USBD controller

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

tmon-nordic commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -0,0 +1,1846 @@
+	USBD_T *const base = config->base;
+
+	/* Set SE0 for disconnect */
+	base->SE0 |= USBD_DRVSE0;
```
This sounds wrong. Is the `SE0` simply some crazy naming to "disable D+ pull-up"? The device cannot initiate bus reset on its own - it is always host responsibility. The Full-Speed device can disable the D+ pull-up which will cause the pull-downs at the host to discharge the lines (and create SE0).

Is the NuMaker USBD publicly available? Does the HW allow disabling D+ pull-up?

ccli8 replied:
It is [publicly available](https://www.nuvoton.com/products/microcontrollers/arm-cortex-m4-mcus/m467-ethernet-crypto-series/?group=Document&tab=2) (check the title M463/M467 Series Technical Reference Manual and the IP is USB 1.1 Device Controller (USBD)).

According to the h/w spec, the effect is to force D+/D- to low, so it can simulate disconnecting device from host. Excerpt the description:

> A software-disconnect function is also supported for this USB controller. It is used to simulate the disconnection of this device from the host. If user enables SE0 bit (USBD_SE0), the USB controller will force the output of USB_D+ and USB_D- to level low and its function is disabled. After disabling the SE0 bit, host will enumerate the USB device again.

tmon-nordic replied:
I wonder if setting `USBD_SE0` is really standard compliant. The specification describes the disconnection as D+ pull-up removal. Anyway, device doesn't have option to disable the pull-up so this is what the software has to do.

ccli8 committed:
soc: nuvoton: numaker: m46x: fix hirc48m typo

Fix typo on HIRC48M. This clock source is required by:
- USB 1.1 OTG PHY
- USBD
- USBH
- OTG

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: change numaker m46x usbd clock source to hirc48m

Change NuMaker M463/M467 series USBD clock source to HIRC48M.
This makes core-clock and its clock source PLL not required
to be multiple of 48MHz.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented:
1. Rebase on main to get device id support #75747 
2. Change usbd clock source to hirc48m so that core clock and PLL are not limited to multiple of 48MHz which usbd requires
3. Fix with end of control data out stage. With #74021 getting in, `net_buf` can over-allocate for `UDC_BUF_GRANULARITY` requirement and now `net_buf_tailroom()` cannot reflect free buffer room exactly as allocate request.

jfischer-no commented:
@ccli8 I see only `boards/nuvoton/numaker_pfm_m467` has USB controller, please add `usbd` test feature for this board.

```diff
diff --git a/boards/nuvoton/numaker_pfm_m467/numaker_pfm_m467.yaml b/boards/nuvoton/numaker_pfm_m467/numaker_pfm_m467.yaml
index d668f51facb..a7e26960ae6 100644
--- a/boards/nuvoton/numaker_pfm_m467/numaker_pfm_m467.yaml
+++ b/boards/nuvoton/numaker_pfm_m467/numaker_pfm_m467.yaml
@@ -13,4 +13,5 @@ ram: 512
 flash: 1024
 supported:
   - gpio
+  - usbd
 vendor: nuvoton
```

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -0,0 +1,1863 @@
+			/* Enable back USB/PHY */
+			base->ATTR |= USBD_ATTR_USBEN_Msk | USBD_ATTR_PHYEN_Msk;
+
+			/* Message for bottom-half processing */
+			msg.type = NUMAKER_USBD_MSG_TYPE_UDC_BUS_EVENT;
+			msg.udc_bus_event.type = UDC_EVT_VBUS_READY;
+			numaker_usbd_send_msg(dev, &msg);
```
This is not necessary for all the bus events, it should be simple
`			udc_submit_event(dev, UDC_EVT_VBUS_READY, 0);`

You only need to offload it to a thread for the end of a transfer and setup packet handling events.

ccli8 replied:
Yes, UDC stack actually can cover bottom-half processing of bus events (like reset device address and un-configure device for bus reset event). Just let UDC stack handle bus events and there is no need to offload for them extra.

ccli8 committed:
boards: nuvoton: numaker_pfm_m467: add usbd

Add usbd test feature for the numaker_pfm_m467 board

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -0,0 +1,1863 @@
+	if (ep_cfg) {
+		ep_cfg->stat.halted = stalled;
+	}
+}
```
Removed `numaker_usbd_ep_get_udc_config` and expanded the function code inline.
> Also, why would ep_cur->addr_valid be false?

Originally, `ep_cur->addr_valid` could be `false` when un-stall all endpoints (including ones without bound addresses) in bus reset handler. With below change, `false` check becomes unnecessary and is removed.

ccli8 commented:
> I see only boards/nuvoton/numaker_pfm_m467 has USB controller, please add usbd test feature for this board.

Added usbd into the numaker_pfm_m467 board.
